misc-utils/fincore.c

```diff
@@ -227,28 +230,36 @@ static int fincore_fd (struct fincore_control *ctl,
  */
 static int fincore_name(struct fincore_control *ctl,
 			const char *name,
-			struct stat *sb,
+			unsigned long long *size,
 			off_t *count_incore)
 {
 	int fd;
 	int rc = 0;
+	struct stat sb;
 
 	if ((fd = open (name, O_RDONLY)) < 0) {
 		warn(_("failed to open: %s"), name);
 		return -errno;
 	}
 
-	if (fstat (fd, sb) < 0) {
+	if (fstat (fd, &sb) < 0) {
```

Comment:
The newly used `blkdev_get_size` actually will fall back to `fstat()` automatically and also ignore `S_ISDIR()`.
I think this can all be reduced to `blkdev_get_size()`.

Reply:
Thank you for the pointer, I looked at the implementation, this would increase the number of ioctls on regular file case, but seems to simplify the code a bit.

Seems like it's hard to tell between genuine failures and not, so I introduced a check for whether files are of expected type, ISREG and ISBLK look like the types that are supported from my understanding.

Reply:
blkdev_get_size interface is a bit weird, does any OS returns the size of block device as a return value to stat? stat usually gives the inode metadata which is usually 0 as a device file doesn't occupy space on disk.

I didn't want to modify blkdev_get_size right now, but might make sense to reorganize so that it actually does a reasonable thing for BLK/CHR devices, and regular files and symlinks